### PR TITLE
#169258643 : Avail requests to manager made by his/her direct reports

### PR DESCRIPTION
**Pull request by ivymwende**
#### What does this PR do?

Created requests should be available to the requester’s manager so that it can be approved. This feature makes that possible.

#### Description of Task to be completed?

- [x]  Add get requests route
- [x]  Add validations
- [x]  Add tests

#### How should this be manually tested?

- Clone this repository using the command: `git clone https://github.com/andela/team-odd-bn-backend.git`
- Checkout to this branch using the command: `git checkout ft-request-approval-169258643`
- Drop all the tables by running the command: `npm run dropTables`
- Add them back by running the command: `npm run db-migrate`
- Then add data some data to the database by running the command: `npm run db-seed-dev`
- Run the application using the command: `npm run dev-start`
- Run the tests using the command: `npm test`

#### Any background context you want to provide?

N/A

#### What are the relevant pivotal tracker stories?
[#169258643](https://www.pivotaltracker.com/story/show/169258643)

#### Screenshots (if appropriate)
Screen Shot 2019-11-28 at 07 04 50
Screen Shot 2019-11-28 at 07 05 27

#### Questions:

### CODE REVIEW

File: src/controllers/TripController.js [hunk 1]
```diff
@@ -69,6 +73,41 @@ class TripController {
       return Customize.errorMessage(req, res, err.message, 500);
     }
   }
-}
 
+  /**
+   * Manager should be able to view approvals
+   * @static
+   * @param {object} req request object
+   * @param {object} res response object
```
Comment (NiyoEric) [inline src/controllers/TripController.js:119]: Add the route in jsDoc to help reviewers remember the api data for example `@description PATCH api/v1/trips/2?status=reject`

File: src/middlewares/Exists.js [hunk 2]
```diff
@@ -25,7 +25,26 @@ class Exists {
...
+    * check is the request exist
+    * @static
+    * @param {object} req request object
+    * @param {object} res response object
+    * @param {object} next next
+    * @memberof Exists
+    * @returns {object} data
+    */
+  static async getLineManager(req, res, next) {
+    // const { managerId } = req.body;
```
Comment (NiyoEric) [inline src/middlewares/Exists.js:60]: please remove comment.

File: src/middlewares/Validate.js [hunk 3]
```diff
@@ -48,6 +48,7 @@ class Validate {
       check('itinerary.*.originId', 'originId should be an integer').isNumeric(),
       check('itinerary.*.destinationId', 'destinationId should be an integer').isNumeric(),
       check('itinerary.*.reason', 'reason should be a minimum of 2 letters').isString().isLength({ min: 2 }),
+      // check('itinerary.*.managerId', 'Invalid Date(format: YYYY-MM-DD)').isAfter(correctDate),
```
Comment (NiyoEric) [inline src/middlewares/Validate.js]: Remove comments everywhere.

File: src/controllers/TripController.js [hunk 4]
```diff
@@ -69,6 +73,41 @@ class TripController {
...
+                }
+              ]
+            }
+          ],
+        }
+      ],
+    });
+    const result = tripRequestsToSpecificManager.map(request => request.user);
+    return Customize.successMessage(req, res, 'Requests fetched successfully', result, 200);
+  }
```
Comment (NiyoEric) [inline src/controllers/TripController.js:136]: The query is complex, I suggest we use **try and catch** technique to handle unexpected server error.

File: src/tests/010-userTest.js [hunk 5]
```diff
@@ -5,18 +5,12 @@ import mockData from './mock/mockData';
...
 chai.should();
-
 let token;
-
-
 const { expect } = chai;
-
 const {
-  usersSignin, unknownUserUpdate, invalidManager, userInvalidImage
+  unknownUserUpdate, invalidManager, userInvalidImage
```
Comment (NiyoEric) [inline src/tests/010-userTest.js]: In order to avoid complexity, I suggest you don't destruct here, remember the mockdata is wrapped into an object, use it to call property of object you want.

**Review by NiyoEric — CHANGES_REQUESTED** [review]
@ivymwende Good job. I left feedback that will help you improving your work

File: src/controllers/TripController.js (hunk 4 above)
Comment (ivymwende) [reply src/controllers/TripController.js:136]: @NiyoEric  Please note I have refactored the code and implemented the feedback to wrap in a try and catch block

File: src/tests/010-userTest.js (hunk 5 above)
Comment (ivymwende) [reply src/tests/010-userTest.js]: @NiyoEric Please explain on what you mean by complexity?
This is because I was trying to follow the ES6 principle of destructuring.

Comment (ivymwende) [reply src/tests/010-userTest.js]: @NiyoEric Thanks I now understand why

File: src/database/models/tripRequests.js [hunk 6]
```diff
@@ -23,7 +23,7 @@ module.exports = (sequelize, DataTypes) => {
        {onUpdate: 'cascade'});
     tripRequests.hasMany(
       models.trips,
-       {foreignKey: 'id'},
+       {targetKey: 'tripRequestId'},
```
Comment (hezronkimutai) [inline src/database/models/tripRequests.js:26]: Why did you remove the foreignkey, you could just add that line.

File: src/middlewares/Validate.js [hunk 7]
```diff
@@ -108,20 +108,6 @@ class Validate {
...
-      check('originId', 'originId should be an integer').isNumeric(),
-      check('destinationId', 'destinationId should be an integer').isNumeric(),
-      check('reason', 'Reason should be a minimum of 2 letters').isString().isLength({ min: 2 }),
-      check('startDate', 'Start date should be a valid date after today(YY-MM-DD) ').isAfter().isISO8601(),
-    ];
-  }
-
   /**
 * Validate input
 * @static
```
Comment (hezronkimutai) [inline src/middlewares/Validate.js:113]: Why did you remove this lines.

File: src/database/seeders/20191115043919-super-admin.js [hunk 8]
```diff
@@ -18,7 +18,7 @@ module.exports = {
       {
         firstName: 'Brian',
         lastName: 'Maiyo',
-        email: 'maiyo.brian@andela.com',
+        email: 'manager.email@gmail.com',
```
Comment (victkarangwa) [inline src/database/seeders/20191115043919-super-admin.js]: Thanks for changing this email

File: src/middlewares/Exists.js [hunk 9]
```diff
@@ -25,7 +25,26 @@ class Exists {
...
+    const { id } = req.user;
+    const userIdProfile = await userProfile.findOne({ where: { userId: id } });
+    if (!userIdProfile.managerId) {
+      return Customize.errorMessage(req, res, 'Please update your line manager', 404);
```
Comment (victkarangwa) [inline src/middlewares/Exists.js:64]: Nice shot!

**Review by victkarangwa — APPROVED** [review]

**Review by victkarangwa — COMMENTED** [review]
Let's find a way to make travis pass

**Review by victkarangwa — APPROVED** [review]

**Review by victkarangwa — CHANGES_REQUESTED** [review]
Good job @nziokaivy 
Let's just find a way to make `continuous-integration/travis-ci/push` pass.

**Review by victkarangwa — APPROVED** [review]

File: src/controllers/TripController.js [hunk 10]
```diff
@@ -69,6 +74,26 @@ class TripController {
...
+   * @returns {object} data
+   * @description PATCH api/v1/trips/requests
+   */
+  static async getRequestsByManager(req, res) {
+    const { id } = req.user;
+    const result = await availtripRequestsToManager(id);
+    try {
+      if (result) {
+        return Customize.successMessage(req, res, 'Requests fetched successfully', result, 200);
+      }
```
Comment (william000000) [inline src/controllers/TripController.js]: Try not to place `if condition` in the controller

**Review by william000000 — CHANGES_REQUESTED** [review]

File: src/controllers/TripController.js (hunk 10 above)
Comment (ivymwende) [reply src/controllers/TripController.js]: @william000000 Please explain this better?

**Review by hezronkimutai — COMMENTED** [review]
Please look at the comments I left.

File: src/middlewares/Validate.js (hunk 7 above)
Comment (ivymwende) [reply src/middlewares/Validate.js:113]: There are existed two of these functions just different naming for both functions so I just checked for the one being used and removed the unused one.

File: src/database/models/tripRequests.js (hunk 6 above)
Comment (ivymwende) [reply src/database/models/tripRequests.js:26]: hasMany or hasOne relationships inserts the association key in the target model by using the target key . In this case, we are inserting the tripRequestId into the trips model. The only difference is that target is abit more specific.

**Review by hezronkimutai — APPROVED** [review]
Nice job Ivy

**Review by NiyoEric — APPROVED** [review]
